File: demo/gallery.R
## gallery.R — render one example panel per R plot type, from synthetic data.
##
##   Rscript demo/gallery.R                 # base theme -> gallery_R/
##   Rscript demo/gallery.R --theme avm
##
## Mirrors demo/gallery.py. Same synthetic shapes, same design decisions —
## the two tracks should be visually interchangeable.

suppressPackageStartupMessages(library(ggplot2))

args <- commandArgs(trailingOnly = TRUE)
theme_name <- if ("--theme" %in% args) args[which(args == "--theme") + 1] else "base"
root <- normalizePath(file.path(dirname(sub("^--file=", "", grep("^--file=",
  commandArgs(FALSE), value = TRUE)[1])), ".."), mustWork = FALSE)
if (!dir.exists(file.path(root, "R"))) root <- getwd()

source(file.path(root, "R", "figkit.R"))
source(file.path(root, "R", "themes_avm.R"))

th <- if (theme_name == "avm") fk_theme_avm() else fk_theme()
fk_set_theme(th)

out <- file.path(root, "gallery_R")
dir.create(out, recursive = TRUE, showWarnings = FALSE)
cat(sprintf("[gallery] theme=%s -> %s\n", theme_name, out))

set.seed(0)

## ── synthetic data (mirrors demo/synth.py) ─────────────────────────────────
n_genes <- 2000
genes <- c(sprintf("GENE%04d", seq_len(n_genes - 41)),
           paste0("MT-ND", 1:6), paste0("RPL", 1:10), paste0("RPS", 1:7),
           paste0("LINC", 1000:1007), paste0("MIR", 100:103),
           "AL100000.1", "AL100001.1",
           "Negative01", "Negative02", "SystemControl1", "FalseCode3")
lfc <- rnorm(n_genes, 0, 0.35)
base_mean <- 10^runif(n_genes, 0.5, 3.5)
real <- sample(n_genes, 60)
lfc[real] <- lfc[real] + sample(c(-1, 1), 60, TRUE) * runif(60, 0.8, 3.2)
se <- 0.25 + 1.5 / sqrt(base_mean)
# 2*pnorm(-|z|) is the accurate upper tail; 1 - pnorm(|z|) underflows to 0.
pval <- pmax(2 * pnorm(-abs(lfc / se)), 1e-300)
de <- data.frame(gene = genes, log2FoldChange = lfc, pvalue = pval,
                 baseMean = base_mean, stringsAsFactors = FALSE)

hallmarks <- c("HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_HYPOXIA",
               "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION",
               "HALLMARK_ANGIOGENESIS", "HALLMARK_INFLAMMATORY_RESPONSE",
               "HALLMARK_KRAS_SIGNALING_UP", "HALLMARK_MYC_TARGETS_V1",
               "HALLMARK_E2F_TARGETS", "HALLMARK_G2M_CHECKPOINT",
               "HALLMARK_INTERFERON_GAMMA_RESPONSE", "HALLMARK_APOPTOSIS",
               "HALLMARK_P53_PATHWAY", "HALLMARK_GLYCOLYSIS",
               "HALLMARK_MTORC1_SIGNALING", "HALLMARK_NOTCH_SIGNALING",
               "HALLMARK_TGF_BETA_SIGNALING", "HALLMARK_COMPLEMENT",
               "HALLMARK_COAGULATION")
nes <- rnorm(length(hallmarks), 0, 1.4)
nes[1:4] <- nes[1:4] + runif(4, 0.8, 1.8)
nes[15:18] <- nes[15:18] - runif(4, 0.8, 1.6)
gsea <- data.frame(pathway = hallmarks, NES = nes,
                   padj = pmax(10^(-abs(nes) * runif(length(nes), 1, 2.6)), 1e-30),
                   stringsAsFactors = FALSE)

tfs <- c("RELA", "NFKB1", "JUN", "FOS", "STAT1", "STAT3", "MYC", "E2F1",
         "SP1", "KLF2", "KLF4", "HIF1A", "EPAS1", "SOX17", "ERG", "FLI1",
         "GATA2", "TCF4", "SMAD3", "TEAD1", "FOXO1", "ETS1", "SRF", "HES1")
delta <- rnorm(length(tfs), 0, 1)
delta[1:5] <- delta[1:5] + runif(5, 1, 2.5)
delta[20:24] <- delta[20:24] - runif(5, 1, 2.2)
tf <- data.frame(source = tfs, delta = delta,
                 q = pmax(10^(-abs(delta) * runif(length(delta), 0.8, 3)), 1e-25),
                 stringsAsFactors = FALSE)

ec <- FK_AVM_AV_ORDER
ct <- names(FK_CELL_TYPE_COLORS)[1:8]
labels <- c(ec, ct)
ang <- seq(0, 2 * pi, length.out = length(labels) + 1)[seq_along(labels)]
centers <- cbind(cos(ang), sin(ang)) * 7
n_per <- sample(200:600, length(labels), TRUE)
umap <- do.call(rbind, lapply(seq_along(labels), function(i) {
  data.frame(umap_1 = rnorm(n_per[i], centers[i, 1], 1.1),
             umap_2 = rnorm(n_per[i], centers[i, 2], 1.1),
             label = labels[i], stringsAsFactors = FALSE)
}))
umap$score <- rnorm(nrow(umap)) + (umap$label %in% FK_EC_VENOUS) * 2.2
umap$is_mutant <- runif(nrow(umap)) < ifelse(umap$score > 1.5, 0.25, 0.02)

pal <- c(FK_EC_SUBTYPE_COLORS, FK_CELL_TYPE_COLORS)
if (theme_name != "avm") pal <- NULL

## ── panels ─────────────────────────────────────────────────────────────────
ok <- 0; total <- 0
# The gallery is a reference, not a print artifact: PNG only at 200dpi keeps it
# a few MB instead of tens. Real panels want fk_save_panel's defaults
# (600dpi + cairo_pdf).
render <- function(name, expr, w, h) {
  total <<- total + 1
  cat(sprintf("[render] %s\n", name))
  tryCatch({
    p <- force(expr)
    fk_save_panel(p, name, out, width = w, height = h, dpi = 200,
                  formats = "png")
    ok <<- ok + 1
  }, error = function(e) cat(sprintf("  [FAIL] %s: %s\n", name, conditionMessage(e))))
}

render("01_volcano", {
  fk_volcano(de, x_col = "log2FoldChange", y_col = "pvalue", label_col = "gene",
             lfc_thresh = 0.8, p_thresh = 0.01, top_n_labels = 12,
             exclude_patterns = c(FK_NOISE_PATTERNS, FK_CONTROL_PATTERNS),
             x_label = "log2FC (proximal - distal)",
             y_label = "-log10 nominal p") +
    ggtitle("Proximal vs distal")
}, 6.5, 5.5)

render("02_dotplot_pathway", {
  fk_dotplot_pathway(gsea, top_n = 16, sort_by = "signed",
                     x_label = "NES (proximal - distal)") +
    ggtitle("Hallmark enrichment")
}, 7, 5.5)

render("04_umap_categorical", {
  fk_umap_categorical(umap, "label", palette = pal, point_size = 0.5,
                      legend_title = "Cell type", legend_ncol = 2) +
    ggtitle("Atlas by cell type")
}, 8, 5)

render("06_umap_highlight", {
  fk_umap_highlight(umap, umap[umap$is_mutant, ], fg_size = 1.1) +
    ggtitle(sprintf("Variant-positive cells (n=%d)", sum(umap$is_mutant)))
}, 6, 5)

render("09_lollipop_tf", {
  fk_lollipop(tf, top_n = 22, y_label = "TF activity Δ (proximal - distal)") +
    ggtitle("Transcription-factor activity")
}, 8, 4)

render("11_ordered_bar", {
  bar <- data.frame(cell_type = ec,
                    pct_variant = pmax(runif(length(ec), 0, 3) +
                                       seq(0, 4, length.out = length(ec)), 0),
                    sig = c("", "", "", "*", "*", "**", "***", "**"),
                    stringsAsFactors = FALSE)
  fk_ordered_bar(bar, value_col = "pct_variant", label_col = "cell_type",
                 order = ec,
                 palette = if (theme_name == "avm") FK_EC_SUBTYPE_COLORS else NULL,
                 sig_col = "sig", x_label = "Variant+ cells (% of subtype)") +
    ggtitle("Variant burden along the arteriovenous axis")
}, 6.5, 3.8)

render("08_heatmap", {
  m <- matrix(rnorm(12 * 8), nrow = 12,
              dimnames = list(fk_prettify_pathway(hallmarks[1:12]), ec))
  m[1:4, 6:8] <- m[1:4, 6:8] + 1.8
  m[9:12, 1:3] <- m[9:12, 1:3] - 1.5
  m[3, 7] <- NA  # exercise the NA-grey path
  fk_heatmap(m, symmetric = TRUE, cbar_label = "signed score") +
    ggtitle("Program activity by subtype")
}, 7, 5)

cat(sprintf("\n[done] %d/%d panels -> %s\n", ok, total, out))
if (ok < total) quit(status = 1)

File: demo/gallery.py
"""Render one example panel per plot type, from synthetic data.

    python demo/gallery.py                # base theme -> gallery/
    python demo/gallery.py --theme avm    # AVM domain theme
    python demo/gallery.py --out /tmp/g   # elsewhere

Every panel here is a working template: find the one that looks like what you
need, and copy its function body. No private data, no network, no HPC paths.
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from figkit import apply_rcparams, quantile_norm, save_panel, set_theme  # noqa: E402
from figkit.plots import (  # noqa: E402
    CONTROL_PROBE_PATTERNS, NOISE_GENE_PATTERNS, add_scale_bar, chord_signed,
    dotplot_expression, dotplot_pathway, embedding_categorical,
    embedding_continuous, heatmap, highlight_mask, lollipop, ordered_bar,
    sankey, stacked_bar, upset, volcano,
)
import synth  # noqa: E402

PANELS = []
FORMATS = ("png",)  # set from --formats
DPI = 200           # set from --dpi


def save(fig, name, out):
    """save_panel with the gallery's format/dpi choice (see --dpi/--formats)."""
    return save_panel(fig, name, out, formats=FORMATS)


def panel(name, description):
    """Register a gallery panel. `description` shows up in the HTML index."""
    def deco(fn):
        PANELS.append((name, description, fn))
        return fn
    return deco


# ── volcano ─────────────────────────────────────────────────────────────────
@panel("volcano", "Differential expression: effect size vs significance, "
                  "colored by direction, with noise-probe filtering.")
def _volcano(out, theme):
    de = synth.de_table()
    fig, ax = plt.subplots(figsize=(6.5, 5.5))
    stats = volcano(
        de, ax, x_col="log2FoldChange", y_col="pvalue", label_col="gene",
        lfc_thresh=0.8, p_thresh=0.01, top_n_labels=12,
        exclude_patterns=NOISE_GENE_PATTERNS + CONTROL_PROBE_PATTERNS,
        x_label="log2FC (proximal − distal)", y_label="−log10 nominal p",
        theme=theme,
    )
    ax.set_title("Proximal vs distal", fontsize=12)
    save(fig, "01_volcano", out)
    return stats


# ── pathway dotplot ─────────────────────────────────────────────────────────
@panel("dotplot_pathway", "Pathway enrichment: x/color = NES, size = FDR. "
                          "sort_by='signed' keeps both directions.")
def _pathway(out, theme):
    gs = synth.pathway_table()
    fig, ax = plt.subplots(figsize=(6.5, 6))
    stats = dotplot_pathway(gs, ax, top_n=18, sort_by="signed", theme=theme,
                            x_label="NES (proximal − distal)")
    ax.set_title("Hallmark enrichment", fontsize=12)
    save(fig, "02_dotplot_pathway", out)
    return stats


# ── expression dotplot ──────────────────────────────────────────────────────
@panel("dotplot_expression", "Gene x group expression: color = mean, "
                             "size = % expressing, blocked by direction.")
def _expr(out, theme):
    de = synth.de_table()
    d = de[de["pvalue"].notna()].copy()
    d["score"] = d["log2FoldChange"].abs() * -np.log10(d["pvalue"].clip(lower=1e-300))
    keep = ~d["gene"].str.contains("MT-|RPL|RPS|LINC|MIR|Negative|SystemControl|FalseCode|^AL",
                                   regex=True)
    d = d[keep]
    up = d[d["log2FoldChange"] > 0].nlargest(10, "score")["gene"].tolist()
    dn = d[d["log2FoldChange"] < 0].nlargest(10, "score")["gene"].tolist()
    genes = up + dn

    mean_df, pct_df = synth.expression_matrices(genes, ("proximal", "distal"))
    fig, ax = plt.subplots(figsize=(8.5, 2.9))
    stats = dotplot_expression(
        mean_df, pct_df, ax, theme=theme,
        group_colors={"proximal": theme.up if theme else None,
                      "distal": theme.down if theme else None},
        gene_colors={**{g: (theme.up if theme else "#c51b8a") for g in up},
                     **{g: (theme.down if theme else "#1f3a6e") for g in dn}},
        dividers=[len(up) - 0.5],
        block_labels=[(len(up) / 2 - 0.5, "proximal-up", theme.up if theme else "#c51b8a"),
                      (len(up) + len(dn) / 2 - 0.5, "distal-up", theme.down if theme else "#1f3a6e")],
        cbar_label="mean log1p expr",
    )
    save(fig, "03_dotplot_expression", out)
    return stats


# ── UMAP: categorical ───────────────────────────────────────────────────────
@panel("umap_categorical", "Embedding colored by category, one scatter per "
                           "level, legend mounted outside the axes.")
def _umap_cat(out, theme):
    xy, obs = synth.embedding()
    palette = None
    if theme is not None and "ec_subtype" in theme.palettes:
        from figkit.themes.avm import CELL_TYPE_COLORS, EC_SUBTYPE_COLORS
        palette = {**EC_SUBTYPE_COLORS, **CELL_TYPE_COLORS}
    fig, ax = plt.subplots(figsize=(7, 5))
    stats = embedding_categorical(ax, xy, obs["label"], palette=palette,
                                  point_size=1.6, legend_title="Cell type",
                                  theme=theme)
    ax.set_title("Atlas by cell type", fontsize=12)
    save(fig, "04_umap_categorical", out)
    return stats


# ── UMAP: continuous ────────────────────────────────────────────────────────
@panel("umap_continuous", "Embedding colored by a continuous score; high "
                          "values drawn last so signal isn't buried.")
def _umap_cont(out, theme):
    xy, obs = synth.embedding()
    fig, ax = plt.subplots(figsize=(6, 5))
    embedding_continuous(ax, xy, obs["score"], point_size=1.8,
                         norm=quantile_norm(obs["score"], 0.02, 0.98),
                         cbar_label="signature score", theme=theme)
    ax.set_title("Signature score", fontsize=12)
    save(fig, "05_umap_continuous", out)
    return {"n_points": len(xy)}


# ── UMAP: two-layer highlight ───────────────────────────────────────────────
@panel("umap_highlight", "Two-layer scatter: grey backdrop of all cells + "
                         "highlighted subset. The backdrop is the control.")
def _umap_hl(out, theme):
    xy, obs = synth.embedding()
    mask = obs["is_mutant"].values
    fig, ax = plt.subplots(figsize=(6, 5))
    highlight_mask(ax, xy, mask, label=f"variant+ (n={int(mask.sum())})",
                   fg_size=6, bg_size=1.2, theme=theme)
    ax.legend(loc="upper right", frameon=False, fontsize=9, markerscale=2)
    ax.set_title("Variant-positive cells on atlas", fontsize=12)
    save(fig, "06_umap_highlight", out)
    return {"n_highlighted": int(mask.sum())}


# ── spatial map ─────────────────────────────────────────────────────────────
@panel("spatial_map", "Tissue map in real units: continuous score + scale bar.")
def _spatial(out, theme):
    xy, obs = synth.spatial()
    fig, ax = plt.subplots(figsize=(6.5, 4.6))
    embedding_continuous(ax, xy, obs["score"], point_size=3.5,
                         norm=quantile_norm(obs["score"], 0.02, 0.98),
                         cbar_label="niche score", theme=theme,
                         x_label="", y_label="")
    add_scale_bar(ax, length=0.5, unit="mm", position="lower right")
    ax.set_title("Spatial niche score", fontsize=12)
    save(fig, "07_spatial_map", out)
    return {"n_points": len(xy)}


# ── heatmap ─────────────────────────────────────────────────────────────────
@panel("heatmap", "Signed score matrix; diverging scale centered on zero, "
                  "NaN cells grey (missing ≠ zero).")
def _heat(out, theme):
    m = synth.score_matrix()
    fig, ax = plt.subplots(figsize=(6.5, 5))
    heatmap(m, ax, symmetric=True, cbar_label="signed score",
            col_label_rotation=45, theme=theme)
    ax.set_title("Program activity by subtype", fontsize=12)
    save(fig, "08_heatmap", out)
    return {"shape": list(m.shape)}


# ── lollipop ────────────────────────────────────────────────────────────────
@panel("lollipop", "Signed per-feature effect (TF activity): length = effect, "
                   "size = −log10 q, star strip = significant.")
def _lolli(out, theme):
    tf = synth.tf_table()
    fig, ax = plt.subplots(figsize=(8, 4))
    stats = lollipop(tf, ax, label_col="source", value_col="delta", q_col="q",
                     top_n=24, y_label="TF activity Δ (proximal − distal)",
                     theme=theme)
    ax.set_title("Transcription-factor activity", fontsize=12)
    save(fig, "09_lollipop_tf", out)
    return stats


# ── stacked composition bar ─────────────────────────────────────────────────
@panel("stacked_bar", "Per-sample composition, normalized to fractions so "
                      "group sizes don't confound composition.")
def _stack(out, theme):
    comp = synth.composition()
    palette = None
    if theme is not None and "ec_subtype" in theme.palettes:
        from figkit.themes.avm import CELL_TYPE_COLORS, EC_SUBTYPE_COLORS
        palette = {**EC_SUBTYPE_COLORS, **CELL_TYPE_COLORS}
    labelled = comp.copy()
    labelled.index = [f"{s} (n={int(n):,})" for s, n in zip(comp.index, comp.sum(axis=1))]
    fig, ax = plt.subplots(figsize=(7.5, 3))
    stats = stacked_bar(labelled, ax, palette=palette, legend_title="Cell type",
                        theme=theme)
    ax.set_title("Composition by sample", fontsize=12)
    save(fig, "10_stacked_bar", out)
    return stats


# ── ordered bar ─────────────────────────────────────────────────────────────
@panel("ordered_bar", "One statistic per category, held in a meaningful order "
                      "(anatomical axis) rather than sorted by value.")
def _obar(out, theme):
    import pandas as pd
    rng = np.random.RandomState(11)
    order = synth.EC_SUBTYPES
    df = pd.DataFrame({
        "cell_type": order,
        "pct_variant": np.clip(rng.uniform(0, 3, len(order)) +
                               np.linspace(0, 4, len(order)), 0, None),
    })
    palette = None
    if theme is not None and "ec_subtype" in theme.palettes:
        palette = theme.palette("ec_subtype")
    fig, ax = plt.subplots(figsize=(6, 3.6))
    stats = ordered_bar(df, ax, value_col="pct_variant", label_col="cell_type",
                        order=order, palette=palette,
                        x_label="Variant+ cells (% of subtype)", theme=theme)
    ax.set_title("Variant burden along the arteriovenous axis", fontsize=12)
    save(fig, "11_ordered_bar", out)
    return stats


# ── chord (optional dep) ────────────────────────────────────────────────────
@panel("chord_signed", "Signed interaction chord: separate bands for increased "
                       "and decreased links. Needs `pycirclize`.")
def _chord(out, theme):
    m = synth.interaction_matrix()
    try:
        chord_signed(m, out / "12_chord_signed", top_n_per_sign=14,
                     title="Ligand–receptor Δ", theme=theme, dpi=DPI, formats=FORMATS)
    except ImportError as e:
        print(f"  [skip] {e}")
        return {"skipped": "pycirclize not installed"}
    return {"n_sectors": m.shape[0]}


# ── upset (optional dep) ────────────────────────────────────────────────────
@panel("upset", "Set intersections across modalities. Needs `upsetplot`.")
def _upset(out, theme):
    sets = synth.gene_sets()
    try:
        upset(sets, out / "13_upset", min_subset_size=1, dpi=DPI, formats=FORMATS)
    except ImportError as e:
        print(f"  [skip] {e}")
        return {"skipped": "upsetplot not installed"}
    return {"n_sets": len(sets)}


# ── sankey ──────────────────────────────────────────────────────────────────
@panel("sankey", "Label transfer: where each query label landed in the "
                 "reference's vocabulary.")
def _sankey(out, theme):
    m = synth.label_transfer()
    fig, ax = plt.subplots(figsize=(7.0, 4.2))
    stats = sankey(m, ax, normalize="source", min_flow=0.02,
                   source_label="Query", target_label="Reference", theme=theme)
    save(fig, "19_sankey", out)
    return stats


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--theme", default="base", choices=["base", "avm"])
    ap.add_argument("--out", default=None, help="output dir (default gallery/)")
    ap.add_argument("--only", default=None, help="render only this panel name")
    ap.add_argument("--dpi", type=int, default=200,
                    help="gallery is a reference, not a print artifact — 200 "
                         "keeps it light. Use 600 for real panels (default: 200)")
    ap.add_argument("--formats", default="png",
                    help="comma-separated, e.g. png,svg (default: png)")
    args = ap.parse_args()
    global FORMATS, DPI
    FORMATS = tuple(f.strip() for f in args.formats.split(",") if f.strip())
    DPI = args.dpi

    theme = None
    if args.theme == "avm":
        from figkit.themes.avm import AVM_THEME
        theme = set_theme(AVM_THEME)
    else:
        from figkit import BASE_THEME
        theme = set_theme(BASE_THEME)

    out = Path(args.out) if args.out else Path(__file__).resolve().parents[1] / "gallery"
    out.mkdir(parents=True, exist_ok=True)
    apply_rcparams(theme)
    # apply_rcparams installs the theme's publication dpi (600). The gallery is
    # a reference, so override it — 600dpi PNGs of 6000-point scatters make the
    # committed gallery an order of magnitude bigger than the code.
    matplotlib.rcParams["savefig.dpi"] = args.dpi

    print(f"[gallery] theme={args.theme} dpi={args.dpi} "
          f"formats={','.join(FORMATS)} -> {out}")
    failures = []
    for name, _desc, fn in PANELS:
        if args.only and args.only != name:
            continue
        print(f"[render] {name}")
        try:
            stats = fn(out, theme)
            if stats:
                print(f"         {stats}")
        except Exception as e:
            import traceback
            traceback.print_exc()
            failures.append((name, repr(e)))

    n = len(PANELS) if not args.only else 1
    print(f"\n[done] {n - len(failures)}/{n} panels -> {out}")
    if failures:
        for name, err in failures:
            print(f"  [FAIL] {name}: {err}")
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
